walkeeper/src/wal_service.rs

```diff
@@ -72,99 +78,99 @@ fn read_into(r: &mut impl Read, buf: &mut BytesMut) -> io::Result<usize> {
     Ok(num_bytes)
 }
 
-/*
- * Unique node identifier used by Paxos
- */
-#[repr(C)]
-#[derive(Debug, Clone, Copy, Ord, PartialOrd, PartialEq, Eq)]
+/// Unique node identifier used by Paxos
+#[derive(Debug, Clone, Copy, Ord, PartialOrd, PartialEq, Eq, Serialize, Deserialize)]
 struct NodeId {
```

Comment:
NodeId should provide the same comparsion result both at C and Rust. It should be treated as (term:u64,uuid:u8[16]) pair. Rust doesn't provide compasison operator for byte array. This is why I have used u128 type with big endian encoding instead.

Reply:
I don't think I understand. If our communication with C code is a TCP socket, then only the serialized form matters, not what Rust uses in memory.

Is a place where we are using FFI to link against a C function directly?

What do you mean by "compare the same"? Do you mean `nodeid_1 > nodeid_2` ?

- The previous serialization looked like this:
```rust

struct NodeId {
    term: u64,
    uuid: u128,
}

impl Serializer for NodeId {
    fn pack(&self, buf: &mut BytesMut) {
        buf.put_u128_le(self.uuid);  // <- little-endian
        buf.put_u64(self.term);      // <- big-endian
    }
}
```

There are a few issues here:
- I don't think it's a good design to have one field transmitted big-endian and one field little-endian. If that's true in the C code, then we should fix it in both places.
- The fields are transmitted in the opposite order that they are declared in the struct, making `#[derive(Serialize)]` not work unless we switch them in the struct definition. We should decide on one field order and make it the same everywhere.

If there are requirements that I missed, then we need a unit test to enforce them. The code in this PR passes all of our current tests...

Reply:
> Rust doesn't provide comparison operator for byte array.

Rust arrays and slices do implement `PartialOrd` and `PartialEq`, so comparisons `<` `>` `==` work as expected.

If you would prefer `uuid` to be `[u8; 16]` then that's easy to change.

Reply:
Sorry, looks like I have made some mistake and concluded that arrays in Rust do not implement PartialOrd.
In this case t is certainly better ti use [u8; 16] instread of u128 with big endian encoding. 
But I still do not understand why you have changed  type of term from u64 to  [u8; 8] and put it after uuid.

Answering you questions:
1. No, wal_acceptor is not using FFI now
2. Yes, communication between C and Rust is done by socket.
3. Comparison of NodeId should provide the same result in C and Rust. As far as we are not defining own comparison function for NodeId and instead of it implicitly generate it by inheriting PartialOrd, order of fields in Rust structure should be the same as in C++, shouldn't it? So it is not only matter of serialization.

So, term's type should be u64  and it should be the first field of NodeId type.
Type of uuid should be changed from u128 to [u8;16]

There is still one potential problem: wal_proposer (C) doesn;t somehow enocide its messages, it is just writing struct in socket as it is. So at big endian machine u64 will be sens in big endian order, at little endian - in opposite order. It is not good. But I do not think that we should try to solve this problem right now. May be in future we will use protobuf (or whatever else) for communication protocol between wal_proposer and wal_acceptor (and not only between them). 
In this case hand written pack/unpack functions are not needed as well as using serde for implicit serialization.

Reply:
> But I still do not understand why you have changed type of term from u64 to [u8; 8] and put it after uuid.

**Field order:**
I changed the order because I was trying to match the previous implementation of `pack` and `unpack`, which put `uuid` first. If the previous `pack`/`unpack` had the wrong order, then I am happy to change them back.

**Type of `term`**:
I can make `term` a `u64` again. I only changed it to try to fix the endian problem. If the C code sends native byte order, then I should get the right result.

> wal_proposer (C) doesn;t somehow enocide its messages, it is just writing struct in socket as it is. So at big endian machine u64 will be sens in big endian order, at little endian - in opposite order. It is not good. But I do not think that we should try to solve this problem right now.

I would like to solve this problem as soon as this PR is finished. Using both big-endian and little-endian encodings makes the code too hard to maintain. Pushing native C structs into a socket means the code will be broken for any struct that requires padding for proper alignment.

I would like to fix the Rust `NodeId` to have the same network format as the C code. After reading your comments, I think this should work:
```rust
struct NodeId {
    term: u64,
    uuid: [u8; 16],
}
```

We can serialize/deserialize the C struct (from a little-endian cpu) using this Rust struct + the `LeSer` trait.

Reply:
Yes, this is exactly what I proposing.